core/blockchain.ex

```diff
@@ -23,8 +26,40 @@ defmodule UltraDark.Blockchain do
 
     Ledger.append_block(block)
     UtxoStore.update_with_transactions(block.transactions)
+	rebalance_difficulty? chain
 
     chain
   end
 
+  def rebalance_difficulty(chain) do
+	last = List.first(chain)
+	beginning = Enum.at(chain, max(length(chain) - @diff_rebalance_offset + 1, 0))
+	avg_spb = (unix_timestamp(last.timestamp) - unix_timestamp(beginning.timestamp)) / @diff_rebalance_offset
+	speed_ratio = @target_blocktime / avg_spb
+	prev = last.difficulty
+
+	# difficulty = log speed_ratio base 16 = log2(speed_ratio) / log2(16)
+	diff = :math.log2(speed_ratio) / 4
+	chain = List.update_at(chain, 0, &(%{&1 | difficulty: diff}))
+
+	blue = "\e[34m"
+	clear = "\e[0m"
+
+	IO.puts "#{blue}difficulty of block#{clear} #{length chain} #{blue}set to#{clear} #{diff} #{blue}from#{clear} #{prev}"
+  end
+
+  # rebalances the difficulty, but only if it's needed
+  def rebalance_difficulty?(chain) do
```

Comment:
this function ends with a "?", it should return a boolean rather than perform a function